File: src/lib/tabs/typing.ts
/**
 * "The human is typing" (spec §5: an agent never takes the active tab while
 * they type): a key that went into an editable element of this window within
 * the last TYPING_GRACE_MS, while the window has keyboard focus. 2 s covers
 * the pause between words without holding the guard up for someone who has
 * stopped to read. Mirrored in src-tauri/src/typing.rs.
 */
export const TYPING_GRACE_MS = 2000;

/**
 * A window also tells Rust it is being typed in (`note_typing`), at most this
 * often: the rule above sees only its own window, and an agent's command for
 * another one would bring that window forward mid-word (tabs-questions Q10).
 * Mirrored in src-tauri/src/typing.rs, which widens its grace by it.
 */
export const TYPING_NOTE_INTERVAL_MS = 500;

export function isEditableTarget(target: EventTarget | null): boolean {
  if (target instanceof HTMLInputElement || target instanceof HTMLTextAreaElement) return true;
  return target instanceof Element && target.closest('[contenteditable]:not([contenteditable="false"])') !== null;
}

const MODIFIER_KEYS = new Set(['Shift', 'Meta', 'Control', 'Alt', 'AltGraph', 'CapsLock', 'Fn']);

type KeyLike = Pick<KeyboardEvent, 'key' | 'metaKey' | 'ctrlKey' | 'target'>;

/** A keystroke that edits text: into an editable element, not a shortcut, not a lone modifier. */
export function countsAsTyping(e: KeyLike): boolean {
  return !e.metaKey && !e.ctrlKey && !MODIFIER_KEYS.has(e.key) && isEditableTarget(e.target);
}

export function isHumanTyping(s: { focused: boolean; lastTypedAt: number; now: number }): boolean {
  return s.focused && s.lastTypedAt > 0 && s.now - s.lastTypedAt < TYPING_GRACE_MS;
}

/**
 * The window's typing clock: `note` every keydown (capture phase), ask
 * `typing()`. `report` tells Rust (`note_typing`): only for a key that counts,
 * only while the window has focus, at most once per TYPING_NOTE_INTERVAL_MS.
 */
export function createTypingTracker(deps: { now(): number; focused(): boolean; report?(): void }) {
  let lastTypedAt = 0;
  let lastReportedAt: number | null = null;
  return {
    note(e: KeyLike): void {
      if (!countsAsTyping(e)) return;
      const now = deps.now();
      lastTypedAt = now;
      if (!deps.report || !deps.focused()) return;
      if (lastReportedAt !== null && now - lastReportedAt < TYPING_NOTE_INTERVAL_MS) return;
      lastReportedAt = now;
      deps.report();
    },
    typing(): boolean {
      return isHumanTyping({ focused: deps.focused(), lastTypedAt, now: deps.now() });
    },
  };
}

File: src-tauri/src/typing.rs
//! "The human is typing", app-wide (tabs-questions Q10).
//!
//! Each window keeps its own typing clock (`src/lib/tabs/typing.ts`), and that
//! is enough inside the window: a command there never takes the tab being
//! typed in. It cannot see a command landing in *another* window — that one
//! would come forward, take key focus mid-word, and the next letters would go
//! into the agent's document. So every window also reports its typing here
//! (`note_typing`, at most once per `TYPING_NOTE_INTERVAL_MS`), and an agent's
//! command that would bring a window other than the typed-in one forward is
//! kept in the background (`ai_socket::dispatch`, `window::reveal_window`).

use std::sync::Mutex;
use std::time::{Duration, Instant};

use tauri::{AppHandle, Manager};

/// Mirror of `TYPING_GRACE_MS` in `src/lib/tabs/typing.ts` (a test keeps them equal).
pub const TYPING_GRACE_MS: u64 = 2000;

/// Mirror of `TYPING_NOTE_INTERVAL_MS` there: a window reports at most this
/// often, so the last report can be up to one interval older than the last
/// key. The grace here is widened by it — on the safe side, like the rule.
pub const TYPING_NOTE_INTERVAL_MS: u64 = 500;

/// The window typed in last, and when it reported.
pub struct TypingClock(Mutex<Option<(String, Instant)>>);

impl TypingClock {
    pub fn new() -> Self {
        Self(Mutex::new(None))
    }

    pub fn note(&self, label: &str, at: Instant) {
        *self.0.lock().unwrap() = Some((label.to_string(), at));
    }

    /// The window the human is typing in at `now`, if any.
    pub fn typing_in(&self, now: Instant) -> Option<String> {
        self.0
            .lock()
            .unwrap()
            .as_ref()
            .filter(|(_, at)| typed_recently(*at, now))
            .map(|(label, _)| label.clone())
    }
}

impl Default for TypingClock {
    fn default() -> Self {
        Self::new()
    }
}

/// Whether a report at `at` still counts as typing at `now`.
pub fn typed_recently(at: Instant, now: Instant) -> bool {
    now.saturating_duration_since(at) < Duration::from_millis(TYPING_GRACE_MS + TYPING_NOTE_INTERVAL_MS)
}

/// Whether an agent may not bring `target` forward while the human types in
/// `typing_in`. `target` `None`: a window still to be built. The typed-in
/// window itself is its own frontend's business — it is already in front, and
/// its per-window rule decides what moves there.
pub fn blocks_raise(typing_in: Option<&str>, target: Option<&str>) -> bool {
    match (typing_in, target) {
        (None, _) => false,
        (Some(typed), Some(target)) => typed != target,
        (Some(_), None) => true,
    }
}

/// `blocks_raise` against the app's clock now. `try_state`: a command can
/// arrive before `setup` is done (CLAUDE.md), and an unknown clock is no typing.
pub fn blocks_raise_now(app: &AppHandle, target: Option<&str>) -> bool {
    app.try_state::<TypingClock>()
        .is_some_and(|clock| blocks_raise(clock.typing_in(Instant::now()).as_deref(), target))
}

/// IPC: the calling window's human typed a key that `typing.ts` counts.
#[tauri::command]
pub async fn note_typing(app: AppHandle, window: tauri::WebviewWindow) -> Result<(), String> {
    if let Some(clock) = app.try_state::<TypingClock>() {
        clock.note(window.label(), Instant::now());
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    const GRACE: Duration = Duration::from_millis(TYPING_GRACE_MS + TYPING_NOTE_INTERVAL_MS);

    #[test]
    fn a_report_counts_for_the_grace_plus_one_note_interval() {
        let at = Instant::now();
        assert!(typed_recently(at, at));
        assert!(typed_recently(at, at + GRACE - Duration::from_millis(1)));
        assert!(!typed_recently(at, at + GRACE));
        assert!(typed_recently(at + Duration::from_millis(5), at), "a clock read before the report still counts");
    }

    #[test]
    fn the_clock_answers_the_last_window_while_its_report_is_fresh() {
        let clock = TypingClock::new();
        let t0 = Instant::now();
        assert_eq!(clock.typing_in(t0), None);
        clock.note("editor-3", t0);
        clock.note("main", t0 + Duration::from_millis(100));
        assert_eq!(clock.typing_in(t0 + Duration::from_millis(200)).as_deref(), Some("main"));
        assert_eq!(clock.typing_in(t0 + Duration::from_millis(100) + GRACE), None);
    }

    #[test]
    fn typing_blocks_every_other_window_and_a_new_one_but_not_its_own() {
        assert!(!blocks_raise(None, Some("editor-7")));
        assert!(!blocks_raise(None, None));
        assert!(blocks_raise(Some("editor-3"), Some("editor-7")));
        assert!(blocks_raise(Some("editor-3"), None), "a new window would take the focus too");
        assert!(!blocks_raise(Some("editor-3"), Some("editor-3")));
    }

    #[test]
    fn the_constants_mirror_typing_ts() {
        let ts = include_str!("../../src/lib/tabs/typing.ts");
        for decl in [
            format!("export const TYPING_GRACE_MS = {TYPING_GRACE_MS};"),
            format!("export const TYPING_NOTE_INTERVAL_MS = {TYPING_NOTE_INTERVAL_MS};"),
        ] {
            assert!(ts.contains(&decl), "typing.ts must declare `{decl}`");
        }
    }
}
